site-scrapers/Walgreens.js

```diff
@@ -0,0 +1,117 @@
+        );
+        await page.waitForSelector("#user_name");
+        await page.type("#user_name", process.env.WALGREENS_EMAIL);
+        await page.waitForTimeout(1000);
+        await page.type("#user_password", process.env.WALGREENS_PASSWORD);
+        await page.waitForSelector("#submit_btn:not([disabled])");
+        await page.click("#submit_btn");
+        await page.waitForNavigation().then(
+            () => {},
+            (err) => page.screenshot({ path: "walgreens.png" }) //if you get here, it's almost definitely because login failed.
+        );
+    }
+
+    const isChallenge = await Promise.race([
+        page.waitForSelector("#radio-security").then(() => true),
+        page.waitForSelector("#pf-acc-signout").then(() => false),
+    ]);
+
+    // SECURITY CHALLENGE
+    if (isChallenge) {
+        await page.click("#radio-security");
+        await page.click("#optionContinue");
+        await page.waitForSelector("#secQues");
+        await page.type("#secQues", process.env.WALGREENS_CHALLENGE);
+        await page.click("#validate_security_answer");
+        await page.waitForNavigation();
+    }
+    const availableLocations = {};
+
+    const uniqueZips = [
+        ...new Set(sites.Walgreens.locations.map((site) => site.zip)),
+    ];
+
+    // SEARCH PAGE
+    for (const zip of uniqueZips) {
+        const { latitude, longitude } = usZips[zip];
+        const todayString = new moment().local().format("YYYY-MM-DD");
+
+        let postResponse = await page.evaluate(
+            (latitude, longitude, todayString) =>
+                new Promise((resolve) => {
+                    jQuery
+                        .ajax({
+                            url:
+                                "https://www.walgreens.com/hcschedulersvc/svc/v2/immunizationLocations/timeslots",
+                            type: "POST",
+                            dataType: "json",
+                            data: JSON.stringify({
+                                position: {
+                                    latitude: latitude,
+                                    longitude: longitude,
+                                },
+                                vaccine: { productId: "" },
+                                appointmentAvailability: {
+                                    startDateTime: todayString,
+                                },
+                                radius: 25,
+                                size: 25,
+                                serviceId: "99",
+                                state: "MA",
+                            }),
+                            contentType: "application/json",
+                        })
+                        .then(
+                            (res) => resolve(res),
+                            (err) => {
+                                console.log(JSON.stringify(err)); // helped with debugging, service throws 404 if no appts found
```

Comment:
In order for this to show in a readable log you need to add a listener like `page.on('console', msg => console.log('PAGE LOG:', msg.text));` since otherwise this just logs in the context of the page that puppeteer managing.

Reply:
Oops, I just removed it since we know to treat errors as no results. Is that OK or do you want me to modify?

Reply:
I'm happy with that! Just more of a note and I found it useful for debugging inside the context of the page.